### PR TITLE
Terminate job implementation for eventlet concurrency backend

### LINKED ISSUE #5209: I can‘t stop a task by its task_id
[2018-12-02 23:53:58,955: INFO/MainProcess] Received task: tasks.add[bb1fe102-c1f9-4361-9370-1129900c0d52]
[2018-12-02 23:54:02,479: INFO/MainProcess] Terminating bb1fe102-c1f9-4361-9370-1129900c0d52 (Signals.SIGTERM)
[2018-12-02 23:54:02,490: ERROR/MainProcess] pidbox command error: NotImplementedError("<class 'celery.concurrency.eventlet.TaskPool'> does not implement kill_job",)
Traceback (most recent call last):
  File "d:\envs\aidcs\lib\site-packages\kombu\pidbox.py", line 101, in dispatch
    reply = handle(method, arguments)
  File "d:\envs\aidcs\lib\site-packages\kombu\pidbox.py", line 122, in handle_cast
    return self.handle(method, arguments)
  File "d:\envs\aidcs\lib\site-packages\kombu\pidbox.py", line 116, in handle
    return self.handlers[method](self.state, **arguments)
  File "d:\envs\aidcs\lib\site-packages\celery\worker\control.py", line 163, in revoke
    request.terminate(state.consumer.pool, signal=signum)
  File "d:\envs\aidcs\lib\site-packages\celery\worker\request.py", line 249, in terminate
    pool.terminate_job(self.worker_pid, signal)
  File "d:\envs\aidcs\lib\site-packages\celery\concurrency\base.py", line 115, in terminate_job
    '{0} does not implement kill_job'.format(type(self)))
NotImplementedError: <class 'celery.concurrency.eventlet.TaskPool'> does not implement kill_job
[2018-12-02 23:55:38,956: INFO/MainProcess] Task tasks.add[bb1fe102-c1f9-4361-9370-1129900c0d52] succeeded in 100.0s: 8

this is my main code:

from celery.app.control import Control
from tasks import add, app

myControl=Control(app)
myControl.revoke(task_id="b11729b0-6272-4527-af9d-dc24c0ad492d", terminate=True)

finally，if i want to look at the state of the task only by task_id (just like above), how .

### PR DESCRIPTION
**Pull request by leonidas97**
# Terminate job implementation for eventlet concurrency backend

## Description

Current implementation of **TaskPool** which is based on **eventlet** library does not support terminating jobs. 
Revoking with `terminate=True` or terminating task raises `NotImplementedError: <class 'celery.concurrency.eventlet.TaskPool'> does not implement kill_job` exception while task continues to run. 

Fixes #5209 #3429 
Fixes only eventlet backend for this issue #4019

### CODE REVIEW

File: celery/concurrency/eventlet.py [hunk 1]
```diff
@@ -119,23 +122,34 @@ def on_stop(self):
...
 
     def shrink(self, n=1):
-        limit = self.limit - n
+        limit = self.limit-n
         self._pool.resize(limit)
         self.limit = limit
 
+    def terminate_job(self, pid, signal=None):
+        if pid in self._pool_map.keys():
+            gt = self._pool_map[pid]
```
Comment (thedrow) [inline celery/concurrency/eventlet.py:149]: I think that `greenlet` is a more meaningful and clear variable name than `gt`.

File: celery/concurrency/eventlet.py [hunk 2]
```diff
@@ -144,3 +158,26 @@ def _get_info(self):
...
+
+    @staticmethod
+    def _make_killable_target(target):
+        def killable_target(*args, **kwargs):
+            try:
+                return target(*args, **kwargs)
+            except GreenletExit:
+                return (False, None, None)
+            except Exception:
+                raise
```
Comment (thedrow) [inline celery/concurrency/eventlet.py:170]: Any other exception would be raised anyway.
Why is this necessary?

File: celery/concurrency/eventlet.py [hunk 3]
```diff
@@ -119,23 +122,34 @@ def on_stop(self):
...
+        self._add_to_pool_map(id(gt), gt)
 
     def grow(self, n=1):
         limit = self.limit + n
         self._pool.resize(limit)
         self.limit = limit
 
     def shrink(self, n=1):
-        limit = self.limit - n
+        limit = self.limit-n
```
Comment (thedrow) [inline celery/concurrency/eventlet.py:143]: Please revert this line.
PEP8 requires spaces between operators.

File: celery/concurrency/eventlet.py [hunk 4]
```diff
@@ -107,8 +109,9 @@ def __init__(self, *args, **kwargs):
 
     def on_start(self):
         self._pool = self.Pool(self.limit)
+        self._pool_map = {}
         signals.eventlet_pool_started.send(sender=self)
-        self._quick_put = self._pool.spawn_n
+        self._quick_put = self._pool.spawn
```
Comment (thedrow) [inline celery/concurrency/eventlet.py:114]: Why did we change `spawn_n` to `spawn`?

File: celery/concurrency/eventlet.py [hunk 5]
```diff
@@ -144,3 +158,26 @@ def _get_info(self):
...
+        self._pool_map[pid] = green_thread
+        green_thread.link(
+            TaskPool._cleanup_after_job_finish,
+            self._pool_map,
+            pid
+        )
+
+    @staticmethod
+    def _cleanup_after_job_finish(gt, pool_map, pid):
+        pool_map.pop(pid)
```
Comment (thedrow) [inline celery/concurrency/eventlet.py:183]:
```suggestion
        del pool_map[pid]
```

`dict.pop()` returns a value we don't need. We should use `del` instead as it will just remove the key.

**Review by thedrow — CHANGES_REQUESTED** [review]
Overall, this looks great.
I do have some rather minor comments and a question.

We also need to document this change.

File: celery/concurrency/eventlet.py (hunk 4 above)
Comment (leonidas97) [reply celery/concurrency/eventlet.py:114]: According to [documentation](http://eventlet.net/doc/modules/greenpool.html#eventlet.greenpool.GreenPool.spawn_n) `spawn_n` returns None while `spawn` returns `greenlet`/`greenthread`.

**Comment by leonidas97** [comment]
I implement all requested changes. 

> We also need to document this change.
I am not sure what I need to do here, only documentation I found that talks about eventlet taskpool in details is [this](https://docs.celeryproject.org/en/master/internals/reference/celery.concurrency.eventlet.html#celery-concurrency-eventlet). Rebuilding documentation should add `terminate_job` method to list?
Am I missing something?

**Comment by thedrow** [comment]
We should update the docs [here](https://docs.celeryproject.org/en/stable/userguide/workers.html?highlight=revoke#revoke-revoking-tasks) to reflect that the eventlet pool also supports cancellation.

**Comment by leonidas97** [comment]
> We should update the docs [here](https://docs.celeryproject.org/en/stable/userguide/workers.html?highlight=revoke#revoke-revoking-tasks) to reflect that the eventlet pool also supports cancellation.

updated.

**Review by auvipy — APPROVED** [review]
